chainkit: expose getblockheader

jharveyb opened the pull request:
## Change Description
Description of change / link to associated issue.

Addresses https://github.com/lightninglabs/taproot-assets/issues/315.

Expose the `GetBlockHeader` call from `ChainKit`. This was already available from each possible `chain.Interface`, but not available for users of `ChainKit`; a similar call was added to `NeutrinoKit`, but that would not work for users not running with a Neutrino backend.

## Steps to Test
Steps for reviewers to follow to test the change.

## Pull Request Checklist
### Testing
- [ ] Your PR passes all CI checks.
- [ ] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [ ] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [ ] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [ ] Any new logging statements use an appropriate subsystem and logging level.
- [ ]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

jharveyb commented:
Not sure why so many protos were changed with this.

Roasbeef commented on lnrpc/signrpc/signer.swagger.json:
```diff
@@ -394,14 +394,11 @@
       "type": "object",
       "properties": {
-        "type_url": {
+        "@type": {
```
The proto check CI step fails: https://github.com/lightningnetwork/lnd/actions/runs/6632798261/job/18019624558?pr=8111

So this shows that things aren't being generated the way things are in CI. I think double check your set of deps, we have some docs around somewhere that state which versions we use.

Roasbeef commented on lnrpc/chainrpc/chain_server.go:
```diff
@@ -293,6 +297,31 @@ func (s *Server) GetBlock(_ context.Context,
 }
 
+// GetBlockHeader returns a block header given the corresponding block hash.
+func (s *Server) GetBlockHeader(_ context.Context,
```
Should be the last commit oforr things to build properly.

Roasbeef commented on lnwallet/mock.go:
```diff
@@ -364,3 +364,9 @@ func (*mockChainIO) GetBlock(blockHash *chainhash.Hash) (*wire.MsgBlock,
 	return nil, nil
 }
+
+func (*mockChainIO) GetBlockHeader(blockHash *chainhash.Hash) (*wire.BlockHeader,
```
Should be in the same commit that the interfaces is expanded so the commit builds.

Roasbeef reviewed (changes requested):
Useful RPC call! Straight forward changes. 

Comments mainly about getting the commits to build, and also seeing why you're generating diff protos.

Roasbeef commented:
If I build using `make rpc`, then I get a pretty sizeable diff, similar to what CI shows, stuff like: 
```diff
-               var err error
-               var annotatedContext context.Context
-               annotatedContext, err = runtime.AnnotateIncomingContext(ctx, mux, req, "/lnrpc.Lightning/AddInvoice", runtime.WithHTTPPathPattern("/v1/invoices"))
+               rctx, err := runtime.AnnotateIncomingContext(ctx, mux, req, "/lnrpc.Lightning/AddInvoice", runtime.WithHTTPPathPattern("/v1/invoices"))
                if err != nil {
                        runtime.HTTPError(ctx, mux, outboundMarshaler, w, req, err)
                        return
                }
-               resp, md, err := local_request_Lightning_AddInvoice_0(annotatedContext, inboundMarshaler, server, req, pathParams)
+               resp, md, err := local_request_Lightning_AddInvoice_0(rctx, inboundMarshaler, server, req, pathParams)
                md.HeaderMD, md.TrailerMD = metadata.Join(md.HeaderMD, stream.Header()), metadata.Join(md.TrailerMD, stream.Trailer())
-               annotatedContext = runtime.NewServerMetadataContext(annotatedContext, md)
+               ctx = runtime.NewServerMetadataContext(ctx, md)
                if err != nil {
-                       runtime.HTTPError(annotatedContext, mux, outboundMarshaler, w, req, err)
+                       runtime.HTTPError(ctx, mux, outboundMarshaler, w, req, err)
                        return
                }
```

jharveyb commented:
> If I build using `make rpc`, then I get a pretty sizeable diff, similar to what CI shows, stuff like:
> 
> ```diff
> -               var err error
> -               var annotatedContext context.Context
> -               annotatedContext, err = runtime.AnnotateIncomingContext(ctx, mux, req, "/lnrpc.Lightning/AddInvoice", runtime.WithHTTPPathPattern("/v1/invoices"))
> +               rctx, err := runtime.AnnotateIncomingContext(ctx, mux, req, "/lnrpc.Lightning/AddInvoice", runtime.WithHTTPPathPattern("/v1/invoices"))
>                 if err != nil {
>                         runtime.HTTPError(ctx, mux, outboundMarshaler, w, req, err)
>                         return
>                 }
> -               resp, md, err := local_request_Lightning_AddInvoice_0(annotatedContext, inboundMarshaler, server, req, pathParams)
> +               resp, md, err := local_request_Lightning_AddInvoice_0(rctx, inboundMarshaler, server, req, pathParams)
>                 md.HeaderMD, md.TrailerMD = metadata.Join(md.HeaderMD, stream.Header()), metadata.Join(md.TrailerMD, stream.Trailer())
> -               annotatedContext = runtime.NewServerMetadataContext(annotatedContext, md)
> +               ctx = runtime.NewServerMetadataContext(ctx, md)
>                 if err != nil {
> -                       runtime.HTTPError(annotatedContext, mux, outboundMarshaler, w, req, err)
> +                       runtime.HTTPError(ctx, mux, outboundMarshaler, w, req, err)
>                         return
>                 }
> ```

Looks like the RPC build step is using a bookworm Docker image, with a newer protobuf compiler?

Swtich to bookworm:

```sh
Digest: sha256:445f34008a77b0b98bf1821bf7ef5e37bb63cc42d22ee7c21cc17041070d134f
Status: Downloaded newer image for golang:1.21.0-alpine
Building protobuf compiler docker image...
#0 building with "default" instance using docker driver

#1 [internal] load .dockerignore
#1 transferring context: 2B done
#1 DONE 0.0s

#2 [internal] load build definition from Dockerfile
#2 transferring dockerfile: 1.26kB done
#2 DONE 0.0s

#3 [internal] load metadata for docker.io/library/golang:1.21.0-bookworm
#3 ...

#4 [auth] library/golang:pull token for registry-1.docker.io
#4 DONE 0.0s

#3 [internal] load metadata for docker.io/library/golang:1.21.0-bookworm
#3 DONE 1.0s

#5 [1/4] FROM docker.io/library/golang:1.21.0-bookworm@sha256:640f192f24852cd875514e18a01977ae87692dca466abed7705ad3c4670b6993
```

Compiler install:

```sh
 #6 3.259 After this operation, 252 MB of additional disk space will be used.
#6 3.259 Get:1 http://deb.debian.org/debian bookworm/main amd64 libicu72 amd64 72.1-3 [9376 kB]
#6 3.372 Get:2 http://deb.debian.org/debian bookworm/main amd64 libxml2 amd64 2.9.14+dfsg-1.3~deb12u1 [687 kB]
#6 3.372 Get:3 http://deb.debian.org/debian bookworm/main amd64 libz3-4 amd64 4.8.12-3.1 [7216 kB]
#6 3.425 Get:4 http://deb.debian.org/debian bookworm/main amd64 libllvm14 amd64 1:14.0.6-12 [21.8 MB]
#6 3.575 Get:5 http://deb.debian.org/debian bookworm/main amd64 libclang-cpp14 amd64 1:14.0.6-12 [11.1 MB]
#6 3.651 Get:6 http://deb.debian.org/debian bookworm/main amd64 clang-format-14 amd64 1:14.0.6-12 [75.0 kB]
#6 3.652 Get:7 http://deb.debian.org/debian bookworm/main amd64 clang-format amd64 1:14.0-55.7~deb12u1 [5256 B]
#6 3.652 Get:8 http://deb.debian.org/debian bookworm/main amd64 zlib1g-dev amd64 1:1.2.13.dfsg-1 [916 kB]
#6 3.659 Get:9 http://deb.debian.org/debian bookworm/main amd64 libprotobuf32 amd64 3.21.12-3 [932 kB]
#6 3.666 Get:10 http://deb.debian.org/debian bookworm/main amd64 libprotobuf-lite32 amd64 3.21.12-3 [261 kB]
#6 3.667 Get:11 http://deb.debian.org/debian bookworm/main amd64 libprotobuf-dev amd64 3.21.12-3 [1283 kB]
#6 3.676 Get:12 http://deb.debian.org/debian bookworm/main amd64 libprotoc32 amd64 3.21.12-3 [829 kB]
#6 3.682 Get:13 http://deb.debian.org/debian bookworm/main amd64 protobuf-compiler amd64 3.21.12-3 [83.9 kB]
#6 3.803 debconf: delaying package configuration, since apt-utils is not installed
#6 3.826 Fetched 54.6 MB in 0s (118 MB/s)
#6 3.840 Selecting previously unselected package libicu72:amd64.
#6 3.840 (Reading database ... 
```

rpc-check target detects incorrect compiler version:

```sh
 Verifying protos.
cd ./lnrpc; ../scripts/check-rest-annotations.sh
if test -n "$(git status --porcelain)"; then echo "Protos not properly formatted or not compiled with v3.4.0"; git status; git diff; exit 1; fi
Protos not properly formatted or not compiled with v3.4.0
```

guggero commented:
Ah, the error message in the `rpc-check` target is out of date. Just run `make rpc`, commit the changes and the `make rpc-check` shouldn't fail (it basically detects a dirty git working directory, so you can only run it after you committed the changes).

jharveyb commented:
> Ah, the error message in the `rpc-check` target is out of date. Just run `make rpc`, commit the changes and the `make rpc-check` shouldn't fail (it basically detects a dirty git working directory, so you can only run it after you committed the changes).

This worked, and the issue with testing before is from not updating the protos and `chainkit.yaml` at the same time.

Added a CLI command and itest + release note change, should be ready now.

guggero commented on lnrpc/chainrpc/chainkit.proto:
```diff
@@ -12,6 +12,11 @@ service ChainKit {
     */
     rpc GetBlock (GetBlockRequest) returns (GetBlockResponse);
 
+    /*
```
Since there is a CLI command now, can add the
```
lncli: `chain getblockheader`
```
annotation for the API docs generator.

guggero commented on cmd/lncli/neutrino_active.go:
```diff
@@ -155,16 +155,16 @@ func isBanned(ctx *cli.Context) error {
 }
 
-var getBlockHeaderCommand = cli.Command{
+var getBlockHeaderNeutrinoCommand = cli.Command{
```
:+1:

guggero commented on docs/release-notes/release-notes-0.17.1.md:
```diff
@@ -36,6 +36,10 @@
 
 ## RPC Additions
+
+* [`chainrpc` `GetBlockHeader`](https://github.com/lightningnetwork/pull/8111)
```
nit: URL is missing `/lnd`, that's why the release notes check complains.

guggero reviewed (commented):
utACK, LGTM (pending proper CI run with acceptable level of green).

jharveyb commented:
Addressed feedback; the latst windows itest failure looked unrelated.

The linter failures are for 3 81-char lines, which are similar to other 81-char lines in those files and IIUC can't cleanly be shortened according to our style guidelines.

Roasbeef commented:
> The linter failures are for 3 81-char lines, which are similar to other 81-char lines in those files and IIUC can't cleanly be shortened according to our style guidelines.

Line ref?

Roasbeef reviewed (approved):
LGTM 🌷


Pending CI (just approved again)

Roasbeef commented on docs/release-notes/release-notes-0.17.1.md:
```diff
@@ -36,6 +36,10 @@
 
 ## RPC Additions
+
+* [`chainrpc` `GetBlockHeader`](https://github.com/lightningnetwork/pull/8111)
```
Needs to properly link to the PR.

jharveyb replied:
Ah snuck out of the rebase, resolved.

guggero commented on contractcourt/channel_arbitrator_test.go:
```diff
@@ -187,6 +187,12 @@ func (*mockChainIO) GetBlock(blockHash *chainhash.Hash) (*wire.MsgBlock, error)
 	return nil, nil
 }
 
+func (*mockChainIO) GetBlockHeader(blockHash *chainhash.Hash) (*wire.BlockHeader,
```
The linter only looks at new code, so that's why this triggers. Can just remove the argument name (`blockHash`) as we're not using the value anyway. Then everything should fit on one line easily:
```
func (*mockChainIO) GetBlockHeader(*chainhash.Hash) (*wire.BlockHeader, error) {
	return nil, nil
}
```

guggero commented on lnwallet/btcwallet/blockchain.go:
```diff
@@ -149,6 +149,15 @@ func (b *BtcWallet) GetBlock(blockHash *chainhash.Hash) (*wire.MsgBlock, error)
+// GetBlockHeader returns a block header for the block with the given hash.
+//
+// This method is a part of the lnwallet.BlockChainIO interface.
+func (b *BtcWallet) GetBlockHeader(blockHash *chainhash.Hash) (*wire.BlockHeader,
```
linter friendly formatting here would be:
```
func (b *BtcWallet) GetBlockHeader(
	blockHash *chainhash.Hash) (*wire.BlockHeader, error) {

	return b.chain.GetBlockHeader(blockHash)
}
```

jharveyb replied:
Thanks, wasn't being creative enough - lint passes locally for me now.

guggero reviewed (approved):
tACK, LGTM :tada:
